### PR TITLE
fix(cli_tools): Integration of `help -h` with `MessageOutput`

### LINKED ISSUE #90: [MessageOutput] fails to integrate with `HelpCommand`
## Premise

Based on the current documentation, it is expected that [`MessageOutput`][mo] redirects _all_ command logs when passed to [`BetterCommandRunner`][bcr]:

- > A proxy for user-provided functions for passing specific log messages.

- > [...] adds additional functionality such as logging, [...]

## Problem

For any program `mock` based on `BetterCommandRunner` with a custom `MessageOutput`, run:

```txt
mock help help
```

or

```txt
mock help -h
```

It is a valid command, as it seeks usage information about the `help` command of the `mock` program. The command log in this case, unfortunately, does not pass through `MessageOutput`.

## Cause

As per my assessment, I believe this problem is caused by the following design:

1. `MessageOutput` is injected through only [`BetterCommand`][bc]

1. `BetterCommand` is a child of upstream [`Command`][c]

1. The private [`HelpCommand`][hc] is a child of `Command` but not `BetterCommand`

1. `BetterCommandRunner` [indirectly adds][hc-add] `HelpCommand` via its [`super`-call][super]

Essentially, in this inheritance schema, `BetterCommand` is a sibling of `HelpCommand` and not a parent. It fails to meet the goal of [`printUsage`][pu].

## Suggestion

I reckon this design bug could be patched by passing the buck to the [upstream package][args] for their private implementation of `HelpCommand`, or, by making their `CommandRunner` an even better `BetterCommandRunner` by choosing `implements` over `extends` such that it not only satisfies all design contracts but also provides the ability to customize the constructor to add a proper `HelpCommand` that inherits from `BetterCommand`.

<sub>

Aside, there is another limitation with both `BetterCommandRunner` and the upstream `CommandRunner` constructor that it fails to allow the customization of its `argParser` similar to [an issue mentioned earlier][misc].

</sub>

[mo]: https://pub.dev/documentation/cli_tools/latest/cli_tools/MessageOutput-class.html
[bcr]: https://pub.dev/documentation/cli_tools/latest/cli_tools/BetterCommandRunner-class.html
[bc]: https://pub.dev/documentation/cli_tools/latest/cli_tools/BetterCommand-class.html
[c]: https://pub.dev/documentation/args/latest/command_runner/Command-class.html
[hc]: https://github.com/dart-lang/core/blob/main/pkgs/args/lib/src/help_command.dart
[hc-add]: https://github.com/dart-lang/core/blob/018e1dc7a7de9511a2d92ac87ebfb36085b8e9e5/pkgs/args/lib/command_runner.dart#L94
[super]: https://github.com/serverpod/cli_tools/blob/4c78b6025993411cc7e520c2314887e117b14cea/packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart#L163
[pu]: https://github.com/dart-lang/core/blob/018e1dc7a7de9511a2d92ac87ebfb36085b8e9e5/pkgs/args/lib/command_runner.dart#L101C8-L101C16
[args]: https://pub.dev/packages/args
[misc]: https://github.com/serverpod/cli_tools/issues/89

### PR DESCRIPTION
**Pull request by indraneel12**
# Resolves #90

Consumers of this package can now  truly expect [`MessageOutput`][mo] to work in the case of usage-text query for [`HelpCommand`][hc] (which does not inherit from `BetterCommand`) even while being tightly-coupled and strictly internal to the upstream dependency `package:args`.

---

## Old Behavior

As shown in the referenced issue, `MessageOutput` fails to work when the usage-text for `help` is solicited by users.

## New Behavior

The desired functionality is achieved, and all tests pass successfully.

### Custom Tests

Screenshot_20251111_010644

### All Tests

Screenshot_20251111_010710

---

#### Features

- Fixes for both `mock help -h` and `mock help help`
- Preserves upstream `HelpCommand` behavior as well
- Non-breaking change

[mo]: https://pub.dev/documentation/cli_tools/latest/cli_tools/MessageOutput-class.html
[hc]: https://github.com/dart-lang/core/blob/20ed966843c5930e00c480e2697b4afb7b3fd8e7/pkgs/args/lib/src/help_command.dart#L10

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrects an edge case so help/usage output for the help command is captured and reported consistently, preserving expected exit behavior.

* **Tests**
  * Adds tests validating help-usage output ordering, logging of usage/exception cases, and parity with upstream behavior for valid and invalid inputs.

* **Documentation**
  * Fixed a typo in internal comments.

### CODE REVIEW

**Comment by indraneel12** [comment]
@coderabbitai Thank you for your insightful review comments, even as they're in the "nitpick" category. While it touches upon genuine concerns, practically, it's not a cause of concern due to the following reasons:

1. As you may notice in the internal (upstream) [implementation][ref] of `HelpCommand`, it strictly returns `null` (implicitly, and explicitly with the `Null` return type for `run`). They use the generic type-parameter `T` in `HelpCommand` only as a means to the end.

1. If `package:args` has a breaking change regarding `help` arguments (extremely unlikely), it'll hopefully not affect end-users because I've attached Test Cases that will fail in this scenario. It shall be noticed by this package's publisher(s) before deployment.

However, I agree there's a potential loophole e.g. in a rare scenario where this package is abandoned and `package:args` has a breaking change for that edge-case. So, I'll update the behavior of that if-block such that it shall not only preserve the validity of corresponding Test Cases but also ensure a better experience for end-users in such a rare scenario.

Thank you for bringing this to my attention.

[ref]: https://github.com/dart-lang/core/blob/20ed966843c5930e00c480e2697b4afb7b3fd8e7/pkgs/args/lib/src/help_command.dart#L25

**Comment by indraneel12** [comment]
@coderabbitai Dear AI, a bit of more repair is necessary to make your PR summary accurate, as this isn't particularly about "[...] usage text immediately" (1) and "bypass command execution and analytics recording" (2).

(1): As per expected behavior, when a user runs `mock help -h` or `mock help help`, the program `mock` should indeed display the usage text for `help` command and exit immediately (this was/is the specification regardless of my PR).

(2): All those stuff are not bypassed, both in normal cases (obviously) and in this edge case (because essentially the entire `BetterCommandRunner.runCommand` is executed, except `CommandRunner.runCommand` but the branch-path taken by this `super` call is substituted and corrected within the neighboring if-block).

This PR is merely about fixing the program behavior in one particular edge-case as described in the referenced issue, without affecting any other behavior/components.

File: packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart [hunk 1]
```diff
@@ -48,6 +48,32 @@ typedef SetLogLevel = void Function({
...
+
+  @override
+  String get description =>
+      'Display help information for ${runner.executableName}.';
+
+  @override
+  String get invocation => '${runner.executableName} $name [command]';
+
+  @override
+  Never run() => throw StateError(
```
Comment (christerswahn) [inline packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart]: Nit: Throw UnimplementedError instead

File: packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart [hunk 2]
```diff
@@ -367,6 +397,36 @@ class BetterCommandRunner<O extends OptionDefinition, T>
     );
   }
 
+  static bool _isUsageOfHelpCommandRequested(final ArgResults topLevelResults) {
```
Comment (christerswahn) [inline packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart]: Suggest moving this to the other file as stated above. Also please add a short doc comment stating the purpose of this function.

File: packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart [hunk 3]
```diff
@@ -48,6 +48,32 @@ typedef SetLogLevel = void Function({
 /// A function type for tracking events.
 typedef OnAnalyticsEvent = void Function(String event);
 
+/// A dummy to replicate the usage-text of upstream private `HelpCommand`
+final class _HelpCommandDummy extends Command {
+  _HelpCommandDummy({required this.runner});
```
Comment (christerswahn) [inline packages/cli_tools/lib/src/better_command_runner/better_command_runner.dart]: This original `help` command is `hidden`, this one should probably be too.

**Review by christerswahn — CHANGES_REQUESTED** [review]
Good contribution! Had some comments and questions.
